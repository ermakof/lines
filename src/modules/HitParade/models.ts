export interface IUserInfo {
  login: string;
  ts?: number;
  score: number;
}

export type TUserList = Record<number, IUserInfo>;
