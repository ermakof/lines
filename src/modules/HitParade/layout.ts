import styled from '@emotion/styled';

export const ColTime = styled.div`
  width: 50%;
`;

export const ColUser = styled.div`
  width: 30%;
`;

export const ColScore = styled.div`
  width: 15%;
`;

export const ColWinner = styled.div`
  width: 10%;
`;

interface IRow {
  winner?: number;
  highlighted?: boolean;
}
export const Row = styled.div<IRow>`
  display: flex;
  line-height: 40px;
  ${({ winner }) => (winner === 1 ? 'color: #ffff00;' : '')}
  ${({ winner }) => (winner === 2 ? 'color: #c0c0c0;' : '')}
  ${({ winner }) => (winner === 3 ? 'color: #a52a2a;' : '')}
  ${({ highlighted }) => (highlighted ? 'background: #aaaaaa45;' : '')}
`;
