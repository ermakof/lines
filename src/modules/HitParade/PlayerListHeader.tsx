import React, { FC, memo } from 'react';
import { ColScore, ColTime, ColUser, ColWinner, Row } from '@src/modules/HitParade/layout';

const PlayerListHeader: FC = () => (
  <Row role="header">
    <ColTime role="col">Время</ColTime>
    <ColUser role="col">Имя</ColUser>
    <ColScore role="col">Счет</ColScore>
    <ColWinner role="col">Место</ColWinner>
  </Row>
);

export default memo(PlayerListHeader);
