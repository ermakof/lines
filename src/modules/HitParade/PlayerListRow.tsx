import React, { FC, memo } from 'react';
import moment from 'moment';

import { IUserInfo } from '@src/modules/HitParade/models';
import { ColScore, ColTime, ColUser, ColWinner, Row } from '@src/modules/HitParade/layout';

export interface IPlayerListRow {
  user: IUserInfo;
  winner: number;
  highlighted: boolean;
}
const PlayerListRow: FC<IPlayerListRow> = ({ user, winner, highlighted = false }) => (
  <Row role="row" winner={winner} highlighted={highlighted}>
    <ColTime role="col">{moment(user.ts).format('DD.MM.YYYY, HH:mm:ss')}</ColTime>
    <ColUser role="col">{user.login}</ColUser>
    <ColScore role="col">{user.score}</ColScore>
    <ColWinner role="col">{winner ? winner : ''}</ColWinner>
  </Row>
);

export default memo(PlayerListRow);
