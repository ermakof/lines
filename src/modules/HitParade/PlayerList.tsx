import React, { FC, memo } from 'react';
import { TUserList } from '@src/modules/HitParade/models';
import Grid from '@src/layout/Grid';
import { sortByScore } from '@src/modules/HitParade/utils';
import PlayerListHeader from '@src/modules/HitParade/PlayerListHeader';
import PlayerListRow from '@src/modules/HitParade/PlayerListRow';

interface IPlayerList {
  list: TUserList;
  loginTime?: number;
}
const PlayerList: FC<IPlayerList> = ({ list, loginTime }) => {
  const userList = Object.values(list).sort(sortByScore);
  return (
    <Grid role="grid">
      <PlayerListHeader />
      {userList.map((user, ind) => {
        return (
          <PlayerListRow
            key={user.ts}
            user={user}
            winner={ind + 1}
            highlighted={user.ts === loginTime}
          />
        );
      })}
    </Grid>
  );
};

export default memo(PlayerList);
