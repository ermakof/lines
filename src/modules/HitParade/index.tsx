/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, memo } from 'react';
import { useSelector } from 'react-redux';

import Title from '@src/layout/Title';
import Grid from '@src/layout/Grid';
import Section from '@src/layout/Section';
import { TRootState } from '@src/store';
import { TUserList } from '@src/modules/HitParade/models';
import { IUserProfile } from '@src/modules/Auth/model/IUserProfile';
import PlayerList from '@src/modules/HitParade/PlayerList';
import styled from '@emotion/styled';

const Root = styled(Section)`
  background: #282c34;
`;

const HitParade: FC = () => {
  const hitParade: TUserList = useSelector(({ app }: TRootState) => app.hitParade);
  const userProfile: IUserProfile = useSelector(({ auth }: TRootState) => auth.userProfile || {});

  return (
    <Root role="panelHitParade">
      <Grid role="rootGrid">
        <Title>Хит-парад</Title>
        {hitParade && <PlayerList list={hitParade} loginTime={userProfile.loginTime} />}
      </Grid>
    </Root>
  );
};

export default memo(HitParade);
