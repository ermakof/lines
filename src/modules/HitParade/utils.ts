import { IUserInfo } from '@src/modules/HitParade/models';

export const sortByScore = (a: IUserInfo, b: IUserInfo) => {
  return a.score > b.score ? -1 : a.score < b.score ? 1 : 0;
};
