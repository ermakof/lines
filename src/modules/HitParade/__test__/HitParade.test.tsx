import React from 'react';
import { useSelector } from 'react-redux';
import { render, screen } from '@testing-library/react';

import HitParade from '@src/modules/HitParade';

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));

const mockSelectors = (selector: any, store: any) => {
  return selector(store);
};

const userProfile = {
  login: 'user',
  password: '123',
  token: '123-123-123-123',
  loginTime: 1646851461537,
};

describe('Hit parade form', () => {
  it('render snapshot', () => {
    const { asFragment } = render(<HitParade />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('render panel without grid', () => {
    const mockStore = {
      app: { hitParade: undefined },
      auth: { userProfile },
    };
    (useSelector as jest.Mock).mockImplementation((selector) => mockSelectors(selector, mockStore));
    render(<HitParade />);
    const panel = screen.getByRole(/panelHitParade/gi);
    expect(panel).toBeInTheDocument();
    const grid = screen.getByRole(/rootGrid/gi);
    expect(grid).toBeInTheDocument();
    const title = screen.getByText(/Хит-парад/gi);
    expect(title).toBeInTheDocument();
  });

  it('render panel with grid', () => {
    const mockStore = {
      app: {
        hitParade: {
          1646851461537: { login: 'Игрок 1', ts: 1646851461537, score: 100 },
          1646855489093: { login: 'Игрок 2', ts: 1646855489093, score: 200 },
          1646859512749: { login: 'Игрок 3', ts: 1646859512749, score: 300 },
        },
      },
      auth: {
        userProfile: {
          login: 'user',
          password: '123',
          token: '123-123-123-123',
          loginTime: 1646851461537,
        },
      },
    };
    (useSelector as jest.Mock).mockImplementation((selector) => mockSelectors(selector, mockStore));
    render(<HitParade />);
    const panel = screen.getByRole(/panelHitParade/gi);
    expect(panel).toBeInTheDocument();
    const grid = screen.getByRole(/rootGrid/gi);
    expect(grid).toBeInTheDocument();
    const title = screen.getByText(/Хит-парад/gi);
    expect(title).toBeInTheDocument();
    const form = screen.getByRole(/hitParade/gi);
    expect(form).toBeInTheDocument();
  });
});
