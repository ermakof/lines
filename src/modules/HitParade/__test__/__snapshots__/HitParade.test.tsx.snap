// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hit parade form render snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1n2h5h0"
    role="panelHitParade"
  >
    <div
      class="css-11wxzwa"
      role="rootGrid"
    >
      <p
        class="css-1890bmp"
      >
        Хит-парад
      </p>
    </div>
  </div>
</DocumentFragment>
`;
